hexrd/instrument/hedm_instrument.py

```diff
@@ -2721,15 +2706,14 @@ def _generate_ring_params(tthr, ptth, peta, eta_edges, delta_eta):
 
     return retas, eta_bins, rtth_idx, reta_idx
 
-
 def _run_histograms(rows, ims, tth_ranges, ring_maps, ring_params, threshold):
     for i_row in range(*rows):
         image = ims[i_row]
 
-        # handle threshold if specified
         if threshold is not None:
+            if sparse.issparse(image):
+                image = image.toarray()
             # !!! NaNs get preserved
-            image = np.array(image)
```

Comment:
I think this was being done to make a new copy of the array so that the original array doesn't get modified by the line below this one.

Are you sure it is okay to modify the original array here?

Reply:
In my testing, it seems that at this point `image` is actually a scipy sparse array, and casting it to a numpy array was important to do the thresholding. It's possible though that this function may be called with different argument types, in which case adding type constraints and checks would be important. I'd like to start adding type annotations to these functions to clear up any confusion about things like this, if that's a change we agree would be useful.

Reply:
@ZackAttack614 I think scipy sparse arrays are actually only used for `.npz` files (and it might only be receiving a sparse array because of the change [here](https://github.com/HEXRD/hexrd/pull/554/files/32cb8418503c231852953a93e0c7214530b7e47c#r1292126145)). We use HDF5 files a lot in some other workflows that involve the GUI.

Joel did perform `np.array(x)` a lot for performing a copy, so I'm still wondering if he might have done this to avoid modifying the original array. I will try to look into this in more detail later.